tools/clang/include/clang/SPIRV/ModuleBuilder.h

```diff
@@ -9,66 +9,84 @@
-    ErrNoActiveFunction,   ///< Tried to finish building non existing function
-    ErrActiveBasicBlock,   ///< Tried to finish building function when there are
-                           ///< active basic block
-    ErrNoActiveBasicBlock, ///< Tried to insert instructions without active
-                           ///< basic block
-  };
-
   /// \brief Constructs a ModuleBuilder with the given SPIR-V context.
   explicit ModuleBuilder(SPIRVContext *);
 
-  /// \brief Begins building a SPIR-V module.
-  Status beginModule();
-  /// \brief Ends building the current module.
-  Status endModule();
-  /// \brief Begins building a SPIR-V function.
-  Status beginFunction(uint32_t funcType, uint32_t returnType);
-  /// \brief Ends building the current function.
-  Status endFunction();
-  /// \brief Begins building a SPIR-V basic block.
-  Status beginBasicBlock();
-  /// \brief Ends building the current SPIR-V basic block with OpReturn.
-  Status endBasicBlockWithReturn();
+  /// \brief Begins building a SPIR-V function. At any time, there can only
+  /// exist at most one function under building. Returns the <result-id> for the
+  /// function on success. Returns zero on failure.
+  uint32_t beginFunction(uint32_t funcType, uint32_t returnType);
+  /// \brief Ends building of the current function. Returns true of success,
+  /// false on failure.
+  bool endFunction();
+
+  /// \brief Creates a SPIR-V basic block. On success, returns the <label-id>
+  /// for the basic block. On failure, returns zero.
+  uint32_t bbCreate();
+  /// \brief Ends building the SPIR-V basic block having the given <label-id>
+  /// with OpReturn. Returns true on success, false on failure.
+  bool bbReturn(uint32_t labelId);
+
+  /// \brief Sets insertion point to the basic block with the given <label-id>.
+  /// Returns true on success, false on failure.
+  bool setInsertPoint(uint32_t labelId);
+
+  inline void requireCapability(spv::Capability);
+
+  inline void setAddressingModel(spv::AddressingModel);
+  inline void setMemoryModel(spv::MemoryModel);
+
+  uint32_t getVoidType();
+  uint32_t getFloatType();
+  uint32_t getVec2Type(uint32_t elemType);
+  uint32_t getVec3Type(uint32_t elemType);
+  uint32_t getVec4Type(uint32_t elemType);
+  uint32_t getFunctionType(uint32_t returnType,
+                           const std::vector<uint32_t> &paramTypes);
 
   /// \brief Takes the SPIR-V module under building.
   std::vector<uint32_t> takeModule();
 
 private:
-  /// \brief Ends building the current basic block.
-  Status endBasicBlock();
-
-  SPIRVContext &theContext;                 ///< The SPIR-V context.
-  SPIRVModule theModule;                    ///< The module under building.
-  llvm::Optional<Function> theFunction;     ///< The function under building.
-  llvm::Optional<BasicBlock> theBasicBlock; ///< The basic block under building.
-  std::vector<uint32_t> constructSite;      ///< InstBuilder construction site.
+  using OrderedBasicBlockMap =
```

Comment:
* Done
* We need to construct basic blocks in a ["readable order"](https://github.com/KhronosGroup/glslang/blob/master/SPIRV/InReadableOrder.cpp) (SPIR-V spec requirement + readability). We can emit basic blocks in the suitable order when doing the traversal in `EmitSPIRVAction` so we need the module builder to remember the order of basic blocks we choose in `EmitSPIRVAction`.